### PR TITLE
Cache - Modification + Filtering

### PR DESCRIPTION
**Pull request by azooz2003-bit**
**TODO:**
- Refactor all models to be compliant with SwiftData + use Strings for ids.
- Define relationships between models.
- Filter for predicate and network req via a keypath and value to compare with.

**Relationships**
- Many-to-one (i.e. Enrollments -> Course): to delete an enrollment once it's removed from a course we can use `@Relationship(deleteRule: .cascade, inverse: \Enrollment.course)`.

https://github.com/user-attachments/assets/36ae75a8-7da9-4806-a4a0-4f3630ca7f3c

### CODE REVIEW

File: CanvasPlusPlayground/Model/Course.swift [hunk 1]
```diff
@@ -280,3 +419,8 @@ struct CalendarLink: Codable, Equatable, Hashable {
     let ics: String
 }
 
+extension String {
```
Comment (rahulon12) [inline CanvasPlusPlayground/Model/Course.swift:422]: Move to separate file?

File: CanvasPlusPlayground/Services/CanvasRepository.swift [hunk 2]
```diff
@@ -56,20 +61,22 @@ struct CanvasRepository {
...
-                throw CacheError.decodingError
-            }
+            let models = try modelContainer.mainContext.fetch(descriptor)
             return models
         }
         
         return models
     }
+    
+    func saveAll() async {
```
Comment (rahulon12) [inline CanvasPlusPlayground/Services/CanvasRepository.swift:75]: `@MainActor`?

**Comment by rahulon12** [comment]
@azooz2003-bit Is this ready to review?

**Comment by azooz2003-bit** [comment]
> @azooz2003-bit Is this ready to review?

As a base implementation yes. I'd still like to verify it with other data.

**Pull request merged**